pkg/experimental/storage/filepath/jsonfile_rest.go

```diff
@@ -0,0 +1,429 @@
+package filepath
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+)
+
+// ErrFileNotExists means the file doesn't actually exist.
+var ErrFileNotExists = fmt.Errorf("file doesn't exist")
+
+// ErrNamespaceNotExists means the directory for the namespace doesn't actually exist.
+var ErrNamespaceNotExists = errors.New("namespace does not exist")
+
+var _ rest.StandardStorage = &filepathREST{}
+
+// NewFilepathREST instantiates a new REST storage.
+func NewFilepathREST(
+	groupResource schema.GroupResource,
+	codec runtime.Codec,
+	rootpath string,
+	isNamespaced bool,
+	newFunc func() runtime.Object,
+	newListFunc func() runtime.Object,
+) rest.Storage {
+	objRoot := filepath.Join(rootpath, groupResource.Group, groupResource.Resource)
+	if err := ensureDir(objRoot); err != nil {
+		panic(fmt.Sprintf("unable to write data dir: %s", err))
+	}
+
+	// file REST
+	rest := &filepathREST{
+		TableConvertor: rest.NewDefaultTableConvertor(groupResource),
+		codec:          codec,
+		objRootPath:    objRoot,
+		isNamespaced:   isNamespaced,
+		newFunc:        newFunc,
+		newListFunc:    newListFunc,
+		watchers:       make(map[int]*jsonWatch, 10),
+	}
+	return rest
+}
+
+type filepathREST struct {
+	rest.TableConvertor
+	codec        runtime.Codec
+	objRootPath  string
+	isNamespaced bool
+
+	muWatchers sync.RWMutex
+	watchers   map[int]*jsonWatch
+
+	newFunc     func() runtime.Object
+	newListFunc func() runtime.Object
+}
+
+var _ rest.Storage = &filepathREST{}
+var _ rest.Creater = &filepathREST{}
+var _ rest.Updater = &filepathREST{}
+var _ rest.GracefulDeleter = &filepathREST{}
+var _ rest.CollectionDeleter = &filepathREST{}
+var _ rest.Getter = &filepathREST{}
+var _ rest.Lister = &filepathREST{}
+var _ rest.Scoper = &filepathREST{}
```

Comment:
from what I can tell, apiregister-gen requires that you implement rest.StandardStorage in order to create the resource verbs. Can we use this interface as a requirement here instead? All the other type checks except Scoper are covered by StandardStorage as well. filepathREST should now support this with the watch implementation.

```suggestion
var _ rest.StandardStorage = &filepathREST{}
var _ rest.Scoper = &filepathREST{}
```

Reply:
> from what I can tell, apiregister-gen requires that you implement rest.StandardStorage in order to create the resource verbs.

i suppose you're talking about `rest.Storage` instead of `rest.StandardStorage`. the former interface is a requirement for customizing REST while `rest.StandardStorage` is just a short-cut composite of `rest.Getter/Lister/Updater/Watcher...` --- it's not a requirement for implementing custom REST for resources, we can e.g. merely implement Getter/Updater interface for a resource so that it only accepts GET/PUT http requests.

and yes, i updated to assert w/ `rest.StandardStorage` as replacement of the scattered assertions using `rest.Getter/Lister/...`.